### Artificial Neyral Network using Keras 
#### Problem Statement :
#### In this section, we have a dataset of sample of customers of a bank with different features related to the customer. The bank has observed that significant amount of the customers have left the bank. With the customer information, our goal is to predict how many more customers are going to leave the bank in the span of next 6 months. This helps the bank to understant where it should concentrate more in order to avoid customer churning.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Read the dataset
df = pd.read_csv('Churn_Modelling.csv')

In [3]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


##### Thus, from the above information, we get to know that our dataset consists of 1000 rows and 14 columns in total. The dataset inlcudes information about various features related to customers as shown above.

In [5]:
X = df.iloc[:, 3 : 13].values
Y = df.iloc[:, 13].values

#### We have two categorical features namely, Gender and Geography. We should process them for accurate modelling results.

#### Handling the Categorical Data

In [6]:
# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:]

#### Split in train and test sets

In [7]:
from sklearn.model_selection import train_test_split
xTrain, xTest, yTrain, yTest = train_test_split(X, Y, test_size=0.25, random_state=0)

#### Perform Sclaing as the data contains wildly unscaled features

In [8]:
from sklearn.preprocessing import StandardScaler
xScaler = StandardScaler()
yScaler = StandardScaler()

xTrain = xScaler.fit_transform(xTrain)
xTest = xScaler.transform(xTest)

#### Now, let's make the Artificial Neural Network

In [9]:
# Import Keras libraries
import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


#### Initialize the ANN

In [10]:
classifier = Sequential()

#### Add Input layer and the First Hidden Layer

In [11]:
classifier.add(Dense(output_dim=6, init='uniform', activation='relu', input_dim=11))

c:\users\sriram veturi\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=11, units=6, kernel_initializer="uniform")`
  """Entry point for launching an IPython kernel.


#### Add the next Hidden Layer

In [12]:
classifier.add(Dense(output_dim=6, init='uniform', activation='relu'))

c:\users\sriram veturi\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", units=6, kernel_initializer="uniform")`
  """Entry point for launching an IPython kernel.


#### Add the output layer. Note that for all the layers except the output layer, we use RELU as activation function but SIGMOID for the output layer as it gives us the probability of the customer leaving the bank. We can then scale the probablity to our desired threshold and get results.

In [13]:
classifier.add(Dense(output_dim=1, init='uniform', activation='sigmoid'))

c:\users\sriram veturi\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="uniform")`
  """Entry point for launching an IPython kernel.


#### Make the predictions with parameters ADAM, BINARY_CROSSENTROPY and ACCURACY.
##### 'ADAM' is the most accurate and efficient type of Stochastic Gradient Descent Algorithm optimizer.
##### 'Binary Cross-Entropy' is used because our output is a binary attribute i.e. whether the customer would stay or not.
##### We update weights using Back Propogation Algorithm for better 'accuracy'.

In [14]:
classifier.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

#### And with the batch size as 10 and the number of epochs as 100, the training starts here...

In [15]:
# Fitting the ANN to training set
classifier.fit(xTrain, yTrain, batch_size=10, nb_epoch=100)

c:\users\sriram veturi\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:2: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  


Epoch 1/100
7500/7500 [==============================] - 1s 101us/step - loss: 0.4895 - acc: 0.7956
Epoch 2/100
7500/7500 [==============================] - 1s 95us/step - loss: 0.4312 - acc: 0.7963
Epoch 3/100
7500/7500 [==============================] - 1s 115us/step - loss: 0.4273 - acc: 0.7963 0s - loss: 0.4398 -
Epoch 4/100
7500/7500 [==============================] - 0s 66us/step - loss: 0.4227 - acc: 0.8045
Epoch 5/100
7500/7500 [==============================] - 0s 63us/step - loss: 0.4188 - acc: 0.8239
Epoch 6/100
7500/7500 [==============================] - 0s 64us/step - loss: 0.4147 - acc: 0.8272
Epoch 7/100
7500/7500 [==============================] - 0s 65us/step - loss: 0.4117 - acc: 0.8319
Epoch 8/100
7500/7500 [==============================] - 0s 64us/step - loss: 0.4094 - acc: 0.8319
Epoch 9/100
7500/7500 [==============================] - 0s 64us/step - loss: 0.4072 - acc: 0.8340
Epoch 10/100
7500/7500 [==============================] - 0s 66us/step - loss: 0.4057 -

7500/7500 [==============================] - 1s 70us/step - loss: 0.3469 - acc: 0.8591
Epoch 83/100
7500/7500 [==============================] - 1s 70us/step - loss: 0.3468 - acc: 0.8619
Epoch 84/100
7500/7500 [==============================] - 1s 71us/step - loss: 0.3460 - acc: 0.8625
Epoch 85/100
7500/7500 [==============================] - 1s 70us/step - loss: 0.3458 - acc: 0.8615
Epoch 86/100
7500/7500 [==============================] - 1s 71us/step - loss: 0.3454 - acc: 0.8607
Epoch 87/100
7500/7500 [==============================] - 1s 70us/step - loss: 0.3458 - acc: 0.8619
Epoch 88/100
7500/7500 [==============================] - 1s 68us/step - loss: 0.3448 - acc: 0.8609
Epoch 89/100
7500/7500 [==============================] - 1s 69us/step - loss: 0.3456 - acc: 0.8625
Epoch 90/100
7500/7500 [==============================] - 1s 68us/step - loss: 0.3450 - acc: 0.8603
Epoch 91/100
7500/7500 [==============================] - 1s 70us/step - loss: 0.3449 - acc: 0.8625
Epoch 92/100


#### The training is done. Let's assume the threshold as 50% and modify the predictions as 1s and 0s.

In [16]:
yPred = classifier.predict(xTest)
yPred = (yPred > 0.5)

#### Confusion Matrix

In [17]:
# Make Confusion Matrix
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_true=yTest, y_pred=yPred)
print(matrix)

[[1877  114]
 [ 241  268]]


#### Obtain the accuracy on test set

In [18]:
# Get accuracy
from sklearn.metrics import accuracy_score
print("Accuracy : ",accuracy_score(yPred, yTest))

Accuracy :  0.858


### The final accuracy of the test set with Artificial Neural Network using Keras is 85.8% which is significantly higher than the other models.